cmd/transcriber/call/live_captions.go

```diff
@@ -0,0 +1,462 @@
+	// This is a little complicated because we might miss a tick (if the transcriber
+	// takes > 1 tick to transcribe). That is why we are keeping prevTranscribedPos.
+	// The goals are:
+	// 1. Clear the window if new (untranscribed) data is silence,
+	//    and silence > removeWindowAfterSilence.
+	// 2. Do not send the window to the transcriber if all new (untranscribed) data is silence.
+
+	removeWindowAfterSilenceSamples := removeWindowAfterSilence.Milliseconds() * trackOutAudioSamplesPerMs
+	prevtranscribedSeg := -1
+	for i, seg := range segments {
+		if prevTranscribedPos >= seg.Start && prevTranscribedPos < seg.End {
+			prevtranscribedSeg = i
+			break
+		}
+	}
+
+	if prevtranscribedSeg == -1 {
+		return false, false
+	}
+
+	for i := prevtranscribedSeg; i < len(segments); i++ {
+		if !segments[i].Silence {
+			return false, false
+		}
+	}
+	silenceLength := segments[len(segments)-1].End - segments[prevtranscribedSeg].Start
+	if silenceLength >= int(removeWindowAfterSilenceSamples) {
+		// 1. untranscribed data is all silence, and there's been enough silence to end this window.
+		return true, true
+	}
+
+	// 2. all new (untranscribed) data is silence, so don't send to the transcriber.
+	return true, false
+}
+
+func cutWindowToSize(trackID string, window []float32, segments []segmentSamples, prevTranscribedPos int) ([]float32, int) {
+	windowGoalSize := int(maxWindowSize.Milliseconds() * trackOutAudioSamplesPerMs)
+
+	for len(window) > windowGoalSize {
+		if len(segments) == 0 {
+			// Should not be possible, but instead of panic-ing, log an error.
+			slog.Error("processLiveCaptionsForTrack: we have zero segments in the window. Should not be possible.",
+				slog.String("trackID", trackID))
+			break
+		}
+
+		var oldestSegment segmentSamples
+		oldestSegment, segments = segments[0], segments[1:]
+		var cutUpTo int
+		if len(segments) == 0 {
+			// We don't have a complete next segment yet: cut to end of oldest segment.
+			cutUpTo = oldestSegment.End
+		} else {
+			// Cut up to start of segment we're keeping.
+			cutUpTo = segments[0].Start
+		}
+		if cutUpTo > len(window) {
+			// Don't panic, defensive, shouldn't happen.
+			cutUpTo = len(window)
+		}
+		window = window[cutUpTo:]
+
+		// Adjust our marker for where we've transcribed.
+		// e.g., prevTranscribedPos was 10, we've cut 6, new pos is 10 - 6 = 4.
+		prevTranscribedPos -= cutUpTo
+	}
+
+	return window, prevTranscribedPos
+}
+
+func (t *Transcriber) startTranscriberPool() {
+	for i := 0; i < t.cfg.LiveCaptionsNumTranscribers; i++ {
+		t.transcriberWg.Add(1)
+		go t.handleTranscriptionRequests(i)
+	}
+}
+
+func (t *Transcriber) handleTranscriptionRequests(num int) {
+	slog.Debug(fmt.Sprintf("live captions, handleTranscriptionRequests: starting transcriber #%d", num))
+
+	transcriber, err := t.newLiveCaptionsTranscriber()
+	if err != nil {
+		slog.Error("live captions, handleTranscriptionRequests: failed to create transcriber",
+			slog.String("err", err.Error()))
+		return
```

Comment:
Is it expected/safe not calling `t.transcriberWg.Done()` in this case?

Reply:
Right, should do that here too, thanks.